### PR TITLE
VTK-m: Use conflict with virtuals variant with ROCm

### PR DESCRIPTION
**Pull request by kwryankrattiger**

### CODE REVIEW

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 1]
```diff
@@ -124,6 +124,7 @@ class VtkM(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("+rocm", when="@:1.6")
     conflicts("+rocm", when="+cuda")
     conflicts("+rocm", when="~kokkos", msg="VTK-m does not support HIP without Kokkos")
+    conflicts("+rocm", when="+virtuals", msg="VTK-m does not support virtual functions with ROCm")
```
Comment (kmorel) [inline var/spack/repos/builtin/packages/vtk-m/package.py:130]: Virtual methods are obsolete and have been completely removed from VTK-m 2.0. I suggest we remove this option altogether.

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 2]
```diff
@@ -251,8 +252,6 @@ def cmake_args(self):
 
             # hip support
             if "+rocm" in spec:
-                options.append("-DVTKm_NO_DEPRECATED_VIRTUAL:BOOL=ON")
```
Comment (kmorel) [inline var/spack/repos/builtin/packages/vtk-m/package.py:254]: I think this option is still necessary for VTK-m 1.9 and before. (The CMake variable will be ignored for VTK-m 2.0 and later.) Perhaps this should be kept in with a comment that it has no effect for VTK-m 2.0 and above.

File: var/spack/repos/builtin/packages/vtk-m/package.py (hunk 1 above)
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/vtk-m/package.py:130]: I think that can be expressed in the variant itself and described in the variant description, then everything else can just be left as is.

File: var/spack/repos/builtin/packages/vtk-m/package.py (hunk 2 above)
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/vtk-m/package.py:254]: It is redundant with the virtuals variant, and is currently a bug. Before this change it was possible to say `vtk-m+rocm+virtuals` which is incorrect. After this change the `virtuals` variant handles the virtual part of this and the `rocm` variant handles the HIP part without ambiguous overlap

Comment (kmorel) [reply var/spack/repos/builtin/packages/vtk-m/package.py:254]: OK. I see now.

**Review by kmorel — APPROVED** [review]
+1

**Review by alalazo — APPROVED** [review]

**Pull request merged**